nemo/collections/tts/modules/submodules.py

```diff
@@ -410,3 +418,254 @@ def forward(self, forward_input: Tuple[torch.Tensor, torch.Tensor]):
+    """
+
+    def __init__(
+        self, reference_encoder, gst_size=128, n_style_token=10, n_style_attn_head=4,
+    ):
+        super(GlobalStyleToken, self).__init__()
+        self.reference_encoder = reference_encoder
+        self.style_attention = StyleAttention(
+            gst_size=gst_size, n_style_token=n_style_token, n_style_attn_head=n_style_attn_head
+        )
+
+    @property
+    def input_types(self):
+        return {
+            "inp": NeuralType(('B', 'D', 'T_spec'), MelSpectrogramType()),
+            "inp_lengths": NeuralType(('B'), LengthsType()),
+        }
+
+    @property
+    def output_types(self):
+        return {
+            "gst": NeuralType(('B', 'D'), EncodedRepresentation()),
+        }
+
+    def forward(self, inp, inp_lengths):
+        style_embedding = self.reference_encoder(inp, inp_lengths)
+        gst = self.style_attention(style_embedding)
+        return gst
+
+
+class ReferenceEncoder(NeuralModule):
+    """
+    Encode mel-spectrograms to an utterance level feature
+    """
+
+    def __init__(self, n_mels, cnn_filters, dropout, gru_hidden, kernel_size, stride, padding, bias):
+        super(ReferenceEncoder, self).__init__()
+        self.filter_size = [1] + list(cnn_filters)
+        self.layers = torch.nn.ModuleList(
+            [
+                Conv2DReLUNorm(
+                    in_channels=int(self.filter_size[i]),
+                    out_channels=int(self.filter_size[i + 1]),
+                    kernel_size=kernel_size,
+                    stride=stride,
+                    padding=padding,
+                    bias=bias,
+                    dropout=dropout,
+                )
+                for i in range(len(cnn_filters))
+            ]
+        )
+        post_conv_height = self.calculate_post_conv_lengths(n_mels, n_convs=len(cnn_filters))
+        self.gru = torch.nn.GRU(
+            input_size=cnn_filters[-1] * post_conv_height, hidden_size=gru_hidden, batch_first=True,
+        )
+
+    @property
+    def input_types(self):
+        return {
+            "inputs": NeuralType(('B', 'D', 'T_spec'), MelSpectrogramType()),
+            "inputs_lengths": NeuralType(('B'), LengthsType()),
+        }
+
+    @property
+    def output_types(self):
+        return {
+            "out": NeuralType(('B', 'D'), EncodedRepresentation()),
+        }
+
+    def forward(self, inputs, inputs_lengths):
+        # BMW -> BWMC (M: mels)
+        x = inputs.transpose(1, 2).unsqueeze(3)
+        x_lens = inputs_lengths
+        x_masks = self.lengths_to_masks(x_lens).unsqueeze(2).unsqueeze(3)
+
+        for layer in self.layers:
+            x = x * x_masks
+            x = layer(x)
+            x_lens = self.calculate_post_conv_lengths(x_lens)
+            x_masks = self.lengths_to_masks(x_lens).unsqueeze(2).unsqueeze(3)
+
+        x = x * x_masks
+        x = x.contiguous().view(x.shape[0], x.shape[1], -1)
+        self.gru.flatten_parameters()
+        _, x = self.gru(x)
```

Comment:
RNN input needs to be padded_sequence (https://pytorch.org/docs/stable/generated/torch.nn.utils.rnn.pack_padded_sequence.html). Simply masking X does not work because the GRU will still receive the 0 entries as inputs.

Reply:
That makes sense! Fixed.